mobile: measure startup time.

janiversen opened the pull request:
<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [ ] Bug fix
- [ ] Functional change
- [? ] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
<!-- Describe your pull request in detail. -->
This is an updated version of the closed PR. 

Even though QML is using a lot of startup time, it is still worth while to see
how other functions behave. Primarily it is important to see how this varies
across devices, and number of dives (large differences).

It is basically similar to debug printf, but is collected in a QString, and never
stored on disk. I assume that some of the log entries will prove not to be significant,
but it is made by the motto "more is better", since we cannot ask testers to do this
many times.

This PR allows us to gather this information for beta versions. Testers should mail
us the log (using the clipboard function) along with type of device and if BT is active.

As requested the functionality is default not active (macros expand to nothing), but
can be activated with the compiler flag -DENABLE_STARTUP_TIMING.

A note on the QML, had the opportunity to get a log from a version where the QML was
compiled with Qt Quick Compiler, startup time was approx. half (on iOS) so this seems 
to be worth while pursuing (maybe we can get the compiler sponsored somehow).

### Changes made:
<!-- Enumerate the changes with 1), 2), 3) etc. -->
<!-- Ensure the test cases are updated if needed. -->
Please see the commits.

### Related issues:
<!-- Reference issues with #<issue-num>. -->
<!-- Write "Fixes #<issue-num" to notify Github that this PR fixes an issue. -->

### Additional information:
<!-- Include sample dive log or other relevant information to allow testing the change where feasible. -->
@dirkhh this patch only makes sense if we make new beta, and ask the testers to send the logs. They can send the logs directly to me, if we want to keep our ML low volume.

### Release note:
<!-- Describe if this change needs a release note present in CHANGELOG.md. -->
<!-- Also, please make sure to add the release note on top of the file CHANGELOG.md. -->

### Documentation change:
<!-- If this PR makes changes to user functionality, then the documentation has to be updated too. -->
<!-- Please, briefly outline here what has changed in terms of the user experience (UX). -->
<!-- If UX changes have been made, a maintainer should apply the 'needs-documentation-change' label. -->

### Mentions:
<!-- Mention users that you want to review your pull request with @<user-name>. Leave empty if not sure. -->
@dirkhh @neolit123 @bstoeger

dirkhh commented:
The failure continues to be caused by network issues at Travis :-(

janiversen commented on subsurface-mobile-main.cpp:
```diff
@@ -19,14 +19,48 @@
+#include "core/ssrf.h"
+#ifdef ENABLE_STARTUP_TIMING
+#include <QElapsedTimer>
+void log_stp(const char *ident, void *buf)
```
well Qstring cannot be used, because it would demand that all cpp sources include QString, which is not guaranteed.

Same goes for the void *.

btw this is a frequently used methid to generalize extern functions and avoid making demands on the including source.

bstoeger replied:
> well Qstring cannot be used, because it would demand that all cpp sources include QString, which is not guaranteed.

The function declaration is in a `#ifdef __cplusplus` block.
I don't see why you couldn't simply put a
```
class QString;
```
forward-declaration in this block. You could even `#include <QString>` in this block.

Perhaps you have to move it out of the `extern "C"` thing. But I don't see the point of declaring a function as `extern "C"`, when it can only be called from C++. And since you return the log as a `QString`, it can only be called from C++. Which again proves that this should be two functions.

Actually, I think it could even be three:
1) Start timer
2) Log
3) Access log

> btw this is a frequently used methid to generalize extern functions and avoid making demands on the including source.

I think opaque pointers were mostly used in days of yore by libraries to avoid applications accessing private data. To prevent disasters in the case that the data format changed. Qt actually does something similar with its PIMPL idiom. But I see no reason why we would do this internally.

dirkhh replied:
ageism... some of us resemble this remark

dirkhh replied:
I don't think there's a point in over-designing this code, but I think there are some good points here.
1) the multi entry to the same function confused me as well and I don't see a strong value in it... the downside of that is the need for a global QMutex, but again, this is just some debugging code, not the cradle of the future of Subsurface
2) returning QString from the function that gets us the string just seems much more reasonable; in that sense the suggestion of 
```
void log_stp(const char *);
QString get_stp_log();
```
seems like a good way to go. Or 
```
void logStp(const char *);
QString getStpLog();
```
Again, this is debug code - naming is, err, not my key concern.
3) I really don't care how the string is assembled - .append is fine by me

janiversen replied:
well 3 functions would mean the variables need to be global, something I was told earlier to avoid. I do not understand the confusion, there are 2 macros calling a common function, a pattern you see often (multiple interface functions calling one implementation function).

making a forward declaration to simulate a #include is at least as ugly as a void *

it is not in the extern “C” thing, it is in the cplusplus part to avoid including it in C code, somethung which wax requested in the earlier reviews.

Design can be discussed for a very long time, I used the best design considering the constraint I was given, clearly if the constraints changes others designs are simpler, but hardly fair, and in this case overengineering do not forget this is the second rework.

Should the 3 fubctions then also be inline so we have extra calls in the release code ? (another constraint change)

bstoeger replied:
Note that I'm only giving suggestions on how I think the code can be improved. I'm not posing constraints. Even without going over-the-top, I think this can be made much more sensible.

> well 3 functions would mean the variables need to be global

No, they're *local* to the translation unit (of static linkage in C++ parlance).

> making a forward declaration to simulate a #include is at least as ugly as a void *

Sorry, here I strongly disagree. Forward-declarations of types to enable strong typing is orders-of-magnitude less ugly than duck-typing via `void *`. You will find forward-declaration is a very common and often recommended idiom.
See for example: Point #7 in http://www.acodersjourney.com/2016/05/top-10-c-header-file-mistakes-and-how-to-fix-them/

> it is not in the extern “C” thing, it is in the cplusplus part to avoid including it in C code

Well, then all the better.

> Should the 3 fubctions then also be inline so we have extra calls in the release code 

I was trying to suggest something different:
```
#ifdef DEBUG
class QString;
extern void log_internal(...); // Note: extern not needed, since default linkage for functions.
extern QString get_log_internal();
inline void log(...)
{
    log_internal(...);
}
inline QString get_log()
{
    return get_log_internal();
}
#else
inline void log(...)
{
}
inline QString get_log()
{
    return QString();
}
#endif
```

Of course, macros work just as well with the well known caveats.

janiversen commented on subsurface-mobile-main.cpp:
```diff
@@ -19,14 +19,48 @@
+					.append(QString::number(stpDuration.elapsed())) \
+					.append(" ms, ") \
+					.append(ident) \
+					.append("\n");
```
no special reason, just my preference.

bstoeger commented on subsurface-mobile-main.cpp:
```diff
@@ -19,14 +19,48 @@
+#include <QElapsedTimer>
+void log_stp(const char *ident, void *buf)
+{
+static bool firstCall = true;
```
The compiler has to do some kind of locking under the hood, because it has to check if this is the first invocation of the function and then construct the objects.

Of course the performance impacts are completely trivial, but I think these objects are fundamentally not function-local.

dirkhh commented:
I will no longer restart the test - this does not look like it is caused by this PR, there seems to be something screwed up with Travis' networking.

bstoeger commented:
Just a final remark:
Since the implementation is in a mobile-only translation unit, shouldn't the declaration also go in a mobile-only header?
On desktop we have `printf`, files, profilers, etc.

dirkhh commented:
> Since the implementation is in a mobile-only translation unit, shouldn't the declaration also go in a mobile-only header?

I don't think we have an obvious "mobile-only" header, do we?
We could add one, sure... but I'm not sure it's worth it.

janiversen commented:
@bstoeger first let me say thank you for your good summary, and then let me explain.

I am sorry if I come over in this PR as stubborn, but the truth is, that I had made a decision when I closed the earlier patch. The intervention from @dirkhh and my respect for ssrf and our community made me try one more time, so I am simply a bit over-sensitive this time. I really like review comments, because they tend to point out things forgotten. Your review comments are mostly based on fact and I have earlier just implemented them. When I get comments that are impossible to implement or "my preference is", that is where discussions start (and in the last PR there was plenty).

I am also used to that the depth of the discussion depend on the impact on ssrf, in this case it is something really inferior, whereas when the new divelist becomes a PR I do expect longer discussions.

Enough of that, now your comments:
1) Your point about only being for MOBILE is implemented (and really something I should have seen)
2) the void * is truly ugly, and replaced with QString * (I guess I overreacted to the comments in the previous PR).
3) add a LOG_STP_START function, not implemented, because it leads to faulty code.
If we add LOG_STP to a global constructor, it may be called before or after main() is called, which
leads to unexpected results in case we have a LOG_STP_START.
using a bool firstCall is a well known design pattern that avoids this kind of racing problem.
4) use of inline functions, not implemented.
If using inline functions, we would need to #ifdef the content of each function. I see the use of
a macro as interface function and behind it an implementation function as a nicer pattern, because 
the macro is voided in release code, and you do not have an empty call, when stepping with the
debugger.
3) split in multiple functions, not implemented.
the design pattern with multiple interface functions and a single implementation function is a well
proven pattern. I believe this is more a matter of personal style than anything else.

I hope my last changes are a balance between the comments received and what I see as important.

janiversen commented:
I agree using Qt before main() is kind of scary, but it can happen with all global classes that uses Qt. The Qt functions (apart from gui) will work, because Qt do not have a special startup call. However please remark I do NOT suggest we use this "feature" I simply want to avoid a potential problem.

Regarding macros and inline function, if you step with the debugger and LOG_STP is not activated, you step into an empty inline function, whereas the macro is jumped over. I guess I explained myself badly in case of macros we have 1 #ifdef surrounding it all, in case of inline functions we need one #ifdef inside each function (because the function itself needs to be defined also in release but the content voided). But again this is a matter of personal taste being more an assembler/C guy I am quite used to use #define for a number of items.

I used to work on a real large project, and they used the interface/implementation pattern, to keep old API´s stable while adding new API´s and keeping a single implementation. That convinced me of the general benefit, but each case is different.

janiversen commented:
@dirkhh can you please have a look at travis, the linux build fails with a git test, something I think is not my cause. Thanks in advance.

I am off in a couple of hours and will not be back until late Sunday (Europe time), I do read mail, but do not have my Mac available.

janiversen commented:
Heh, sometimes time solves problems :-) Travis is green.

janmulder commented:
I kicked the built. That gitstorage test is sometimes confused.

janiversen commented:
@janmulder thanks !

@bstoeger just a question to be sure, are you OK (on purpose not writing happy) with the current version (you have added a review, so you need to approve) ?

bstoeger commented:
I think the interface is really ugly (both, the macros and the internal function), but I don't oppose this change, because there are bigger fish to fry.

In any case in the current version, there's still two points which are definitely not OK:

1) Cast of `QString *` to `QString *`
2) Indentation of the static function-local variables.

Moreover, I would change the `if ... else` into two `if`s. Not only is the interface strange, you also silently impose that exactly one of the arguments is non-null (two times null: crash; one times null: second argument ignored). This calls for a comment.

> you have added a review, so you need to approve

Nope - my review is pretty much irrelevant.

janiversen commented:
@dirkhh I updated the PR with the suggestions from @bstoeger, however I did it online through github so I could not do a “rebase -i” (which I am getting quite good at using, thanks to you), and I not sure about tabs/space (I used a tab). I

dirkhh commented:
I’m at a conference - I can rebase / clean up the commits tonight, but not for about 12 hours... I saw that Derek complained about a missing SOB - I assume that I’m allowed to add that to commits where it is missing?

dirkhh commented:
Yes, I meant collaborator

dirkhh commented:
I ended up collapsing this into one commit, fixing some white space, fixing a compile error.
Usually I don't collapse PRs, but I couldn't figure out how to create "authentic" yet "consistent" commits that reflected your work @janiversen - my apologies.

janiversen commented:
Thanks a lot, and no needs for apologies 1 commit or 5 commits does not matter, the funcionality matters.
